clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -366,15 +391,6 @@ static void collectSYCLAttributes(Sema &S, FunctionDecl *FD,
       FD->dropAttr<SYCLIntelUseStallEnableClustersAttr>();
     }
   }
-
-  // Attributes that should not be propagated from device functions to a kernel.
-  if (DirectlyCalled) {
-    llvm::copy_if(FD->getAttrs(), std::back_inserter(Attrs), [](Attr *A) {
-      return isa<SYCLIntelLoopFuseAttr, SYCLIntelFPGAMaxConcurrencyAttr,
```

Comment:
Shouldn't these be added to the DirectlyCalled list in L362?

Reply:
This attributes currently apply on kernel functor/lambda  in SYCL2020 modes, so i did not add them in L362.
SYCLIntelFPGAMaxConcurrencyAttr,
SYCLIntelFPGADisableLoopPipeliningAttr, 
SYCLIntelFPGAInitiationIntervalAttr

Reply:
These attributes are available prior to SYCL2020 right? Shouldn't they apply for earlier versions as well? I think this patch changes existing behavior for these attributes.

Reply:
> These attributes are available prior to SYCL2020 right? Shouldn't they apply for earlier versions as well? I think this patch changes existing behavior for these attributes.

All attributes were added recently. they were added after SYCl2020 spec release. I think they should not apply in SYCL2017 modes.

https://github.com/intel/llvm/pull/3388
https://github.com/intel/llvm/pull/3441

Reply:
I'm not sure how its expected to work but unless the extensions these attributes support are limited to SYCL2020, or these attributes are documented to work only in SYCL 2020, we probably should not be changing this behavior for earlier versions of SYCL.  @AaronBallman please let us know your thoughts here.

Reply:
Also, if these attributes are not supported for SYCL 2017, shouldn't we be diagnosing it? In the PRs you shared above, I don't see these diagnostics. Some of the tests don't have the standard specified either.

Reply:
>>Also, if these attributes are not supported for SYCL 2017, shouldn't we be diagnosing it?

The diagnostic seems reasonable to me.

Reply:
Please wait for @AaronBallman's input. In my opinion it is confusing/strange to have individual attributes behave differently in different versions of SYCL spec.

Reply:
>>In my opinion it is confusing/strange to have individual attributes behave differently in different versions of SYCL spec, 

I agree with you, @elizabethandrews.

Reply:
Sorry about the delayed response -- having power issues at the house.

> I'm not sure how its expected to work but unless the extensions these attributes support are limited to SYCL2020, or these attributes are documented to work only in SYCL 2020, we probably should not be changing this behavior for earlier versions of SYCL. @AaronBallman please let us know your thoughts here.

Agreed. I think the route we want to go is:
  * When a new attribute is added to the `intel` namespace, we support it in all the language modes where it has valid semantics. The semantics of the attribute can do whatever is most sensible for that given language mode, but once the semantics are set and the attribute has been released in the wild, the semantics should not change except to become more permissive (e.g., we shouldn't change the semantics such that code breaks, but we should be fine to allow the attribute to be used in ways that used to produce an error).
  * When a new attribute is added to the `sycl` namespace, we support it in all the language modes where it has valid semantics, but we diagnose use of a new feature in the older modes as an extension. The semantics of the attribute have to follow what's specified by the SYCL spec. If we think some semantics are going to cause implementation concerns for us, we need to talk to the SYCL spec authors about how to resolve it on a case-by-case basis.
  ** Note: community typically also adds a compatibility warning in the newer mode so people who want their code to remain compatible with older language standards can do so. If we didn't support SYCL 1.2.1, we could skip this diagnostic, but from talks with @kbsmith-intel, it sounds like SYCL 1.2.1 support is still mandatory and so the future compat warning should also be added.
  * When the attribute does not have valid semantics in a given mode (regardless of what vendor namespace the attribute is in), we should ignore the attribute with a diagnostic to let the user know it's being ignored. The only exception to this rule are SYCL attributes that are ignored in host mode but not device mode.

> I guess the question is more - should we change existing behavior for these attributes in SYCL 2017

I don't think we should change existing behaviors -- that runs too much risk of silently breaking user code. However, it's also not clear to much just how much implementation effort it is to retain the old behavior in each case and whether the old behavior was sensible or not in SYCL 1.2.1. My reading of the 1.2.1 spec suggests that *only* the `vec_type_hint`, `work_group_size_hint`, and `reqd_work_group_size` attributes are propagated and none of the rest of them are. I get this from (emphasis added by me for clarity):

> The vec_type_hint, work_group_size_hint and reqd_work_group_size kernel attributes in OpenCL C
apply to kernel functions, but this is not syntactically possible in SYCL. In SYCL, **these attributes** are legal on
device functions and **their** specification is propagated down to any caller of those device functions, such that the
kernel attributes are the sum of all the kernel attributes of all device functions called.

That said, I have no idea if this is an accurate understanding of the SYCL spec.

Reply:
Thanks @AaronBallman and @elizabethandrews. I have added the attributes below in SYCL 1.2.1. The semantic is same in all modes.

SYCLIntelFPGAMaxConcurrencyAttr,
SYCLIntelFPGADisableLoopPipeliningAttr,
SYCLIntelFPGAInitiationIntervalAttr